xtm/xtm_api.cc

```diff
@@ -0,0 +1,282 @@
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		void (*fun)(void *);
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *data;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int readfd;
+	/**
+	 * The file descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int writefd;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+/**
+ * Function gets num data pointers, previously saved in data field of
+ * union xtm_msg. Used in case send/recv xtm API pattern.
+ * @param[in] queue - xtm queue
+ * @param[out] data - array to save data pointers from xtm_msg union data field
+ * @param[in] num - count of elements to get
+ * @retval return the number of data pointers actually read.
+ */
+static inline unsigned
+xtm_queue_get_data(struct xtm_queue *xtm_queue,
+		   void **data, unsigned num)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		if (new_read == queue_write)
+			break;
+		data[i] = queue->buffer[new_read].data;
+		new_read = (new_read + 1) & (queue->len - 1);
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+/**
+ * Function executes all functions contained in all queue elements.
+ * Used in case dispatch/invoke xtm API pattern
+ * @param[in] queue - xtm queue
+ * @retval return count of executed functions
+ */
+static inline unsigned
+xtm_queue_execute_funs(struct xtm_queue *xtm_queue)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for(;;) {
+		if (new_read == queue_write)
+			break;
+		queue->buffer[new_read].fun(queue->buffer[new_read].fun_arg);
+		new_read = (new_read + 1) & (queue->len - 1);
+		i++;
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+static inline int
+xtm_queue_flush_pipe(struct xtm_queue *xtm_queue)
+{
+	unsigned char tmp[XTM_PIPE_SIZE];
+	ssize_t read_bytes;
+	while ((read_bytes = read(xtm_queue->readfd, tmp, sizeof(tmp))) < 0 &&
+	        errno == EINTR)
```

Comment:
I don't really understand what is happening here.

Reply:
Producer thread write 8 bytes to pipe to notify consumer thread. Consumer thread need to read this 8 (or more in case producer thread have time to write several messages) bytes from pipe. Here we done it. Also read function can fail with errno == EINTR in case we catch signal, in this case we need to retry read.

Reply:
What does libev? I don't see any reads in it. Only writes. It also uses pipes. See third_party/libev/ev.c in Tarantool.

Reply:
see `pipecb` function in libev

Reply:
Because we can catch signal and in this case we need to try read again. As you can see we read in the loop, only when read return < 0 and errno is set to EINTR